### ISSUE TITLE
Error loading flib 0.14

### ISSUE THREAD

**Issue by have-fun-was-taken** [have-fun-was-taken/yafc-ce#83]
The error appears only when trying to load flib `0.14`. It doesn't appear on flib `0.13`.
A minimal example is Factorio with only two mods in the `mod-list.json`: base, and flib. Both mods are enabled.
Steps to reproduce:
1. Make sure that you have the `flib` mod enabled, of the version `0.14.0`.
2. Launch YAFC 0.6.2.
3. Get an error when loading the mod `flib`, no matter on the existing or new project.

![grafik](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/f875f059-8782-4841-a26a-2c1ed38f8a52)

I suggest we prioritize fixing that bug because `flib` is a core library, so there's a high chance that some mods will sooner or later require `flib >= 0.14.0`.

**Comment by have-fun-was-taken** [comment]
Writing it here just in case that YAFC 0.6.2 loads fine on flib 0.13.

**Comment by Dorus** [comment]
I've tried some things, and it seems to randomly fail at this lua line:

```lua
  default_graphical_set = {
    base = { filename = data_util.dark_red_button_tileset, position = { 0, 0 }, corner_size = 8 },
    shadow = _ENV.default_dirt,
  },
```
Replacing `data_util.dark_red_button_tileset` with some other tileset at least make it parse this line. It then still fails elsewhere.

My first hunch is that since factorio itself does load, something is wrong with how we parse lua. I checked and the lua dll is 52, while the latest version seems to be 5.4.6.

Do we know what version Factorio uses? Would it be possible to update our lua dll and test that?

**Comment by Dorus** [comment]
Ah nevermind, well the above might still be worth a shot, but i checked and it seems factorio expect lua 5.2.1 and that's probably what's already used (i cannot find the minor version number we have installed, but that shouldn't be relevant).

**Comment by veger** [comment]
Could it be that we load stuff on another order?

Since `data_util` is `nil` it makes sense that `dark_red_button_tileset` cannot be referenced. But the question (for me at least) is why is `data_util` not available to start with? And what (file) is responsible to make `data_util` available?

**Comment by Dorus** [comment]
Been starting for a bit at the changeset for flib from 1.3 to 1.4 https://github.com/factoriolib/flib/compare/f3b14c9cab2264574dbeac60f39f329c8df0f5e5...5db69e545c55e2004012dd252b13068d17657bbe

In many places, `require("__flib__/data-util")` changed to `require("__flib__.data-util")`.
The lines
```lua
if ... ~= "__flib__.data-util" then
  return require("__flib__.data-util")
end
```
was added to many files.

**Comment by veger** [comment]
Indeed, something made them to change the 'require syntax' in https://github.com/factoriolib/flib/commit/85ed0a99fb22dbf903f742da5b62749cdba1c585 and add 'require guards'.
Might be the reason that data-utils is not loaded by YAFC anymore, as the guard will fail and the require is not activated.

~I am no Lua expert and have no idea why to change the `/` into a `.`~ It [is the same](https://stackoverflow.com/questions/65168192/whats-different-in-lua-require-path-with-or) apparently, must be a personal preference of the flib dev.

So it must be the 'require guards' that throw off YAFC? :thinking: 
(I have no idea what the idea behind those is)

**Comment by Dorus** [comment]
Yes this seems to be the issue. Semi speculation from my side, but so far i think this is what happens:

1. We do not properly populate the arguments when calling files from require. When loading an file, it should be loaded with the filename as the first argument. Thus the argument list (refered with `...` in lua) is empty.

   https://www.lua.org/manual/5.3/manual.html#pdf-require
   (was going to put citiation here but cannot find the relevant doc anymore) 
2. flib added a guard on top of all import files using `if ... ~= "__flib__.data-util" then`. Their goal is simple: optimize module loading, lua will load every file only once _for each name used_, thing is you can load the same file in many ways, using `.` or `/`, using full path or partial etc. etc. and lua could end up loading the same file under many names. (flib is used in many mods so this optimization makes some sense). So the guard assures the file is only loaded once.

   https://www.lua.org/pil/8.1.html
   > Therefore, if you load the same file with two different virtual names, it will be loaded twice.
3. but since we did not populate the `...` arguments properly, the guard instead loads nothing.

The solution would be to properly populate the argument when loading a new file. Somewhere in the Require function ( LuaContext.cs:286 ) or the Exec function ( LuaContext.cs:383 )

**Comment by Dorus** [comment]
I tested the above Pr and it indeed does solve at least half the problem.

@SWeini tested with vanilla+flib+milestones and it worked
I tested with  RateCalculator + stlib 1.4 (+ PipeVisualizer) and it crashed on RateCalculator.

After future analyze i now believe YaFC doens't have the above mentioned inefficiency of loading the same file twice, since we store previous loaded files by their filename, not the requested name. This means flib is first loaded with the wrong name, and the guard bails out, then it's loaded again with the correct name, but the cached guard bailout is reused.

I tried to fix the above with some ugly hacks but haven't gotten it to work yet.
